rollout/controller.go

```diff
@@ -376,6 +366,272 @@ func (c *Controller) syncHandler(key string) error {
+				Service: activeSvc,
+				Type:    validation.ActiveService,
+			})
+		}
+		if rollout.Spec.Strategy.BlueGreen.PreviewService != "" {
+			previewSvc, err := c.servicesLister.Services(rollout.Namespace).Get(rollout.Spec.Strategy.BlueGreen.PreviewService)
+			if k8serrors.IsNotFound(err) {
+				fldPath := validation.GetServiceWithTypeFieldPath(validation.PreviewService)
+				return nil, field.Invalid(fldPath, rollout.Spec.Strategy.BlueGreen.PreviewService, err.Error())
+			}
+			if err != nil {
+				return nil, err
+			}
+			services = append(services, validation.ServiceWithType{
+				Service: previewSvc,
+				Type:    validation.PreviewService,
+			})
+		}
+	} else if rollout.Spec.Strategy.Canary != nil {
+		if rollout.Spec.Strategy.Canary.StableService != "" {
+			stableSvc, err := c.servicesLister.Services(rollout.Namespace).Get(rollout.Spec.Strategy.Canary.StableService)
+			if k8serrors.IsNotFound(err) {
+				fldPath := validation.GetServiceWithTypeFieldPath(validation.StableService)
+				return nil, field.Invalid(fldPath, rollout.Spec.Strategy.Canary.StableService, err.Error())
+			}
+			if err != nil {
+				return nil, err
+			}
+			services = append(services, validation.ServiceWithType{
+				Service: stableSvc,
+				Type:    validation.StableService,
+			})
+		}
+		if rollout.Spec.Strategy.Canary.CanaryService != "" {
+			canarySvc, err := c.servicesLister.Services(rollout.Namespace).Get(rollout.Spec.Strategy.Canary.CanaryService)
+			if k8serrors.IsNotFound(err) {
+				fldPath := validation.GetServiceWithTypeFieldPath(validation.CanaryService)
+				return nil, field.Invalid(fldPath, rollout.Spec.Strategy.Canary.CanaryService, err.Error())
+			}
+			if err != nil {
+				return nil, err
+			}
+			services = append(services, validation.ServiceWithType{
+				Service: canarySvc,
+				Type:    validation.CanaryService,
+			})
+		}
+	}
+	return &services, nil
+}
+
+func (c *Controller) getReferencedRolloutAnalyses(rollout *v1alpha1.Rollout) (*[]validation.AnalysisTemplateWithType, error) {
+	analysisTemplates := []validation.AnalysisTemplateWithType{}
+	if rollout.Spec.Strategy.BlueGreen != nil {
+		blueGreen := rollout.Spec.Strategy.BlueGreen
+		if blueGreen.PrePromotionAnalysis != nil {
+			// CanaryStepIndex will be ignored
+			templates, err := c.getReferencedAnalysisTemplates(rollout, blueGreen.PrePromotionAnalysis, validation.PrePromotionAnalysis, 0)
+			if err != nil {
+				return nil, err
+			}
+			analysisTemplates = append(analysisTemplates, templates...)
+		}
+
+		if blueGreen.PostPromotionAnalysis != nil {
+			// CanaryStepIndex will be ignored
+			templates, err := c.getReferencedAnalysisTemplates(rollout, blueGreen.PostPromotionAnalysis, validation.PostPromotionAnalysis, 0)
+			if err != nil {
+				return nil, err
+			}
+			analysisTemplates = append(analysisTemplates, templates...)
+		}
+	} else if rollout.Spec.Strategy.Canary != nil {
+		canary := rollout.Spec.Strategy.Canary
```

Comment:
It's probably a good idea to add a comment about why we don't need to do anything with the background analysis